lib/internal/inspector/_inspect.js

```diff
@@ -44,88 +77,72 @@ class StartupError extends Error {
+        `Timeout (${timeout}) waiting for ${host}:${port} to be free`);
+    }
+    const error = await new Promise((resolve) => {
       const socket = net.connect(port, host);
-      let didRetry = false;
-      function retry() {
-        if (!didRetry && !didTimeOut) {
-          didRetry = true;
-          setTimeout(pingPort, retryDelay);
-        }
-      }
-
-      socket.on('error', (error) => {
-        if (error.code === 'ECONNREFUSED') {
-          resolve();
-        } else {
-          retry();
-        }
-      });
-      socket.on('connect', () => {
-        socket.destroy();
-        retry();
-      });
+      socket.on('error', resolve);
+      socket.on('connect', resolve);
+    });
+    if (error?.code === 'ECONNREFUSED') {
+      return;
     }
-    pingPort();
-  });
+  }
 }
 
-function runScript(script, scriptArgs, inspectHost, inspectPort, childPrint) {
-  return portIsFree(inspectHost, inspectPort)
-    .then(() => {
-      return new Promise((resolve) => {
-        const needDebugBrk = process.version.match(/^v(6|7)\./);
-        const args = (needDebugBrk ?
-          ['--inspect', `--debug-brk=${inspectPort}`] :
-          [`--inspect-brk=${inspectPort}`])
-          .concat([script], scriptArgs);
-        const child = spawn(process.execPath, args);
-        child.stdout.setEncoding('utf8');
-        child.stderr.setEncoding('utf8');
-        child.stdout.on('data', (chunk) => childPrint(chunk, 'stdout'));
-        child.stderr.on('data', (chunk) => childPrint(chunk, 'stderr'));
-
-        let output = '';
-        function waitForListenHint(text) {
-          output += text;
-          if (/Debugger listening on ws:\/\/\[?(.+?)\]?:(\d+)\//.test(output)) {
-            const host = RegExp.$1;
-            const port = Number.parseInt(RegExp.$2);
-            child.stderr.removeListener('data', waitForListenHint);
-            resolve([child, port, host]);
-          }
-        }
-
-        child.stderr.on('data', waitForListenHint);
-      });
-    });
+const debugRegex = /Debugger listening on ws:\/\/\[?(.+?)\]?:(\d+)\//;
+async function runScript(script, scriptArgs, inspectHost, inspectPort,
+                         childPrint) {
+  await portIsFree(inspectHost, inspectPort);
+  const args = ArrayPrototypeConcat(
+    [`--inspect-brk=${inspectPort}`, script],
+    scriptArgs);
+  const child = spawn(process.execPath, args);
+  child.stdout.setEncoding('utf8');
+  child.stderr.setEncoding('utf8');
+  child.stdout.on('data', (chunk) => childPrint(chunk, 'stdout'));
+  child.stderr.on('data', (chunk) => childPrint(chunk, 'stderr'));
+
+  let output = '';
+  return new Promise((resolve) => {
+    function waitForListenHint(text) {
+      output += text;
+      const debug = RegExpPrototypeSymbolMatch(debugRegex, output);
+      if (debug) {
+        const host = debug[1];
+        const port = Number(debug[2]);
+        child.stderr.removeListener('data', waitForListenHint);
+        resolve([child, port, host]);
```

Comment:
If `debug` is off the Promise here will never resolve. Is that what we want?

Reply:
Yes, it will be pending until another `data` event is emitted by `child.stderr` which will make `output` match the expected regex.